gh-127655: Ensure `_SelectorSocketTransport.writelines` pauses the protocol if needed (#127656)
python/cpython: Misc/NEWS.d/next/Security/2024-12-05-21-35-19.gh-issue-127655.xpPoOf.rst

> Fixed the :mod:`asyncio` selector transport not pausing the protocol when the buffer reaches the high water mark.

Author: ```suggestion
Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark.
```

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark.

Reviewer: Is there a reason this doesn't mention writelines?

Author: I adjusted it to include `writelines`

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark.

Author: ```suggestion
Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using `writelines`.
```

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using `writelines`.

Author: ```suggestion
Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using ``writelines``.
```

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using ``writelines``.

Reviewer 2: If it's possible (but not sure), would you be willing to add a link to `writelines` so that users know which "writelines" method you're talking about? (is it https://docs.python.org/3/library/asyncio-protocol.html#asyncio.WriteTransport.writelines, namely <code>:meth:\`asyncio.WriteTransport.writelines\`</code>?)

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using ``writelines``.

Author: ```suggestion
Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.StreamWriter.writelines`.
```

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.StreamWriter.writelines`.

Reviewer 2: And... I think I've commented too fast. Is it the streamwriter's writelines or the asyncio.WriteTransport.writelines? (I think it's the latter...)

Reviewer 2: (I don't have access to my IDE now so I don't which class is actually the correct one)

Author: Its only `_SelectorTransport` https://github.com/python/cpython/blob/69bb1c6c41c9840e4ae9537ff3e8e6eef718a0a8/Lib/asyncio/selector_events.py#L759

Author: I think `WriteTransport` makes the most sense in this context given its been qualified with `selector`

Reviewer 2: That's what I thought.

Author: Tweaked it a bit. I tried to make it more clear that the problem only happens with ``asyncio.SelectorEventLoop``

Author: <img width="821" alt="Screenshot 2024-12-05 at 7 57 17 PM" src="https://github.com/user-attachments/assets/474d698e-461d-4ffb-a816-d5091d2eab69">

RTD preview looks ok to me
https://cpython-previews--127656.org.readthedocs.build/en/127656/whatsnew/changelog.html#security

Author: If some other wording combination makes better sense, please don't hesitate to adjust.

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.StreamWriter.writelines`.

Author: ```suggestion
Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.WriteTransport.writelines`.
```

> Fixed the :mod:`asyncio` selector transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.WriteTransport.writelines`.

Author: ```suggestion
Fixed the :class:`asyncio.SelectorEventLoop` transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.WriteTransport.writelines`.
```

> Fixed the :class:`asyncio.SelectorEventLoop` transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.WriteTransport.writelines`.

Reviewer 3: I suggest to use the actual name `_SelectorSocketTransport`, the current wording sounds like it would covers the SSL impl too which it isn't.

> Fixed the :class:`asyncio.SelectorEventLoop` transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.WriteTransport.writelines`.

Reviewer 3: ```suggestion
Fixed the :class:`!asyncio.selector_events._SelectorSocketTransport` transport not pausing writes for the protocol when the buffer reaches the high water mark when using :meth:`asyncio.WriteTransport.writelines`.
```
